README.md

```diff
@@ -151,7 +149,38 @@ nativeFactTypes:
 
 Note: You only need to specify the verb for the canonical for of the fact type, any synonymous forms will automatically be remapped to the canonical form
 
-
-### Tests
+## Tests
 
 Tests can be found under the `test/` folder, to run the whole suite use `npm test`
+
+## Storing files and other large objects
+
+An application can choose between two types to save file content or another large object: `File` or `WebResource`. When using a `File`, PineJS saves the content in the database using a binary data type like `BYTEA` or `BLOB`. When using a `WebResource`, PineJS saves the binary content on an external storage service and then writes metadata, including the content public URL, to the database. Client apps use the `WebResource` `href` to get the content.
+
+
+### WebResource
+
+Type [`WebResource`](./src//types/web-resource.ts) can be used to persist files or other large content on an external object storage service like Amazon S3. By "object storage" we refer to a service that can store the content and provide a URL to access that content.
+
+In order to save a `WebResource` you send an instance of [`WebResourceInput`](./src/types/web-resource.ts#L18)
+
+```js
+{
+  filename: string;
+  data: Buffer;
+  contentType?: string;
+  contentDisposition?: string;
+  size?: number;
+  storage: string;
+}
+```
+
+A typical use case is to have a web app where users can upload a file. The app will use [multer](https://github.com/expressjs/multer) to get the file from the http request as a `Buffer` and pass it as the `data` attribute of a `WebResourceInput`.
+
+When retrieving a `WebResource` the `data` attribute is replaced by an `href`( URL ) attribute. See [`WebResourceRef`](./src/types/web-resource.ts#L7)
+
+#### Storage Adapters
+
+The `storage` attribute specifies the name of an [`StorageAdapter`](./src/storage-adapters/storage-adapter.ts#L7). A `StorageAdapter` saves the content to a specific storage, performing a function similar to what a database driver provides. For example, [pinejs-s3-storage](https://github.com/balena-io-modules/pinejs-s3-storage) saves the content to S3 and returns a URL to the persisted object. For testing purposes, this module uses [disk-storage-adapter](./test/storage-adapters/disk-storage-adapter.js).
```

Comment:
Could https://github.com/balena-io-modules/pinejs-s3-storage be a minIO storage driver as well?
It's my very own opinion but using minIO as an example storage driver is opening the usage spectrum. Narrowed S3 implementation may be needed but even for testing we will end up with CI executed minIO, right?

Reply:
We're using minio, see for example https://github.com/balena-io-modules/pinejs-s3-storage/blob/master/docker-compose-local.yaml#L5

I think we will be using our own `https://github.com/balena-io/open-balena-s3`; I plan to replace minio with this one.